### PR TITLE
feat(terraform): Assign/override main vertices by the first new vertice.

### PR DESCRIPTION
**Pull request by ChanochShayner**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

## Description
Assign/override main vertices by the first new vertice to avoid mismatch in the vertices indexes.

## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my feature, policy, or fix is effective and works
- [x] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/terraform/graph_builder/foreach_handler.py [hunk 1]
```diff
@@ -216,21 +224,21 @@ def _create_new_resource(self, main_resource: TerraformBlock, new_value: int | s
...
+            self.local_graph.vertices.append(new_resource)
 
-    def _create_new_resources_foreach(self, statement: list[str] | dict[str, Any], main_resource: TerraformBlock) -> None:
+    def _create_new_resources_foreach(self, statement: list[str] | dict[str, Any], block_idx: int) -> None:
+        main_resource = self.local_graph.vertices[block_idx]
         if isinstance(statement, list):
-            for new_value in statement:
-                self._create_new_resource(main_resource, new_value, new_key=new_value)
+            for i, new_value in enumerate(statement):
+                resource_idx = block_idx if i == 0 else None
```
Comment (tronxd) [inline checkov/terraform/graph_builder/foreach_handler.py:236]:
```suggestion
                resource_idx = block_idx if i == 0 else -1
```
Let's keep indices with their original type

**Review by tronxd — APPROVED** [review]

File: checkov/terraform/graph_builder/foreach_handler.py [hunk 2]
```diff
@@ -216,21 +224,21 @@ def _create_new_resource(self, main_resource: TerraformBlock, new_value: int | s
 
         idx_to_change = new_key or new_value
         self._add_index_to_block_properties(new_resource, idx_to_change)
-        self.local_graph.vertices.append(new_resource)
+        if resource_idx is not None:
+            self.local_graph.vertices[resource_idx] = new_resource
```
Comment (bo156) [inline checkov/terraform/graph_builder/foreach_handler.py:228]: This logic is a bit scary in my opinion, as it basically relies on the fact that you need to pass the correct input to this function in order for it to work, meaning sending `resoucre_idx = None` or not pass it at all.
Instead you can move line 161 to be just here, so the logic of assigning `None` to `resource_idx` will be part of the function itself and not separated from it.

File: checkov/terraform/graph_builder/foreach_handler.py (hunk 1 above)
Comment (bo156) [inline checkov/terraform/graph_builder/foreach_handler.py:236]: seems like you are also doing the same here, so moving this logic like I said earlier will actually save you this line as well

File: checkov/terraform/graph_builder/foreach_handler.py [hunk 3]
```diff
@@ -216,21 +224,21 @@ def _create_new_resource(self, main_resource: TerraformBlock, new_value: int | s
...
-            for new_key, new_value in statement.items():
-                self._create_new_resource(main_resource, new_value, new_key=new_key)
-
-    def _delete_main_resource(self, block_index_to_statement: FOR_EACH_BLOCK_TYPE) -> None:
-        new_vertices = [
-            block for i, block in enumerate(self.local_graph.vertices) if i not in block_index_to_statement or not block_index_to_statement[i]
-        ]
-        self.local_graph.vertices = new_vertices
+            for i, (new_key, new_value) in enumerate(statement.items()):
+                resource_idx = block_idx if i == 0 else None
```
Comment (bo156) [inline checkov/terraform/graph_builder/foreach_handler.py:240]: again

**Review by bo156 — CHANGES_REQUESTED** [review]

File: checkov/terraform/graph_builder/foreach_handler.py [hunk 4]
```diff
@@ -200,7 +201,14 @@ def _build_key_to_val_changes(new_val: str, new_key: str):
             EACH_KEY: new_key
         }
 
-    def _create_new_resource(self, main_resource: TerraformBlock, new_value: int | str, new_key: Optional[str] = None):
+    def _create_new_resource(
+            self,
+            main_resource: TerraformBlock,
```
Comment (YaaraVerner) [inline checkov/terraform/graph_builder/foreach_handler.py:206]: you don't need to get the main_resource as a parameter, you are getting the `resource_idx` so you can get it directly from the graph

**Review by bo156 — APPROVED** [review]

**Pull request merged**